core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +118,135 @@ where
+        new_cache.push(root);
+        self.cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
-        self,
-        mut index: usize,
-        mut merkle_path: Option<&mut Vec<H256>>,
+        &self,
+        mut end_index: usize,
+        mut start_path: Option<&mut Vec<H256>>,
+        mut end_path: Option<&mut Vec<H256>>,
     ) -> H256 {
-        assert!(index < self.hashes.len(), "invalid tree leaf index");
-
         let depth = tree_depth_by_size(self.binary_tree_size);
-        if let Some(merkle_path) = merkle_path.as_deref_mut() {
-            merkle_path.reserve(depth);
+        if let Some(left_path) = start_path.as_deref_mut() {
+            left_path.reserve(depth);
+        }
+        if let Some(right_path) = end_path.as_deref_mut() {
+            right_path.reserve(depth);
         }
 
-        let mut hashes = self.hashes;
-        let mut level_len = hashes.len();
+        let mut hashes = self.hashes.clone();
+        let mut start_index = self.start_index;
+
         for level in 0..depth {
             let empty_hash_at_level = self.hasher.empty_subtree_hash(level);
-
-            if let Some(merkle_path) = merkle_path.as_deref_mut() {
-                let adjacent_idx = index ^ 1;
-                let adjacent_hash = if adjacent_idx < level_len {
-                    hashes[adjacent_idx]
-                } else {
-                    empty_hash_at_level
-                };
-                merkle_path.push(adjacent_hash);
+            if start_index % 2 == 1 {
+                hashes.push_front(self.cache[level]);
+            }
+            if hashes.len() % 2 == 1 {
+                hashes.push_back(empty_hash_at_level);
             }
 
-            for i in 0..(level_len / 2) {
+            let push_sibling_hash = |path: Option<&mut Vec<H256>>, index: usize| {
+                // `index` is relative to `head_index`
+                if let Some(path) = path {
+                    let sibling = ((start_index + index) ^ 1) - start_index + start_index % 2;
+                    // When trimming all existing leaves,
+                    // the first untrimmed leaf and its sibling might not exist.
+                    let hash = hashes.get(sibling).copied().unwrap_or_default();
+                    path.push(hash);
+                }
+            };
+
+            push_sibling_hash(start_path.as_deref_mut(), 0);
+            push_sibling_hash(end_path.as_deref_mut(), end_index);
+
+            let level_len = hashes.len() / 2;
+            for i in 0..level_len {
                 hashes[i] = self.hasher.compress(&hashes[2 * i], &hashes[2 * i + 1]);
             }
-            if level_len % 2 == 1 {
-                hashes[level_len / 2] = self
-                    .hasher
-                    .compress(&hashes[level_len - 1], &empty_hash_at_level);
-            }
 
-            index /= 2;
-            level_len = level_len / 2 + level_len % 2;
+            hashes.drain(level_len..);
```

Comment:
Nit: IIUC, you can use [`VecDeque::truncate()`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.truncate) here and in similar places.